### PR TITLE
Speed up safe_strtod and safe_strtof functions by using double-conversion library

### PR DESCRIPTION
**Pull request by akindyakov**
#11713

### CODE REVIEW

**Comment by AKindyakov** [comment]
@rmlarsen, performance improvement is so attractive :)
There is benchmark: https://gist.github.com/AKindyakov/2c528beccb11a24c1e6ec02ec12df47e
DC is significantly faster.

**Comment by rmlarsen** [comment]
@AKindyakov thanks. That is indeed a nice speedup. 
@martinwicke @vrv what are your opinions? Should we introduce this new dependency or try to speed up decode_csv in another way?

**Comment by rmlarsen** [comment]
We will investigate and see what direction to take.

**Comment by drpngx** [comment]
@rmlarsen any decision on this?

I think we might be open sourcing the google library, which is probably faster than the TF re-implementation.

**Comment by rmlarsen** [comment]
@drpngx I have not had time to pursue this.

**Comment by AKindyakov** [comment]
JFYI patch about case insensitivity for special values was accepted and merged into the master branch.
https://github.com/google/double-conversion/pull/47

**Comment by drpngx** [comment]
Oh, I hadn't realized this was from google. @martinwicke wdyt about taking on this additional dependency?

**Comment by martinwicke** [comment]
I have no problem with it. Seems lightweight enough. We should double-check whether it supports all platforms, but I see MSVC files there, so I presume the answer is yes.

File: third_party/double-conversion/add_case_insensitive_nan.patch
Comment (drpngx) [inline third_party/double-conversion/add_case_insensitive_nan.patch]: Please undo, as per comment, case insensitivity should already be there.

File: tensorflow/core/lib/strings/str_util.cc
Comment (drpngx) [inline tensorflow/core/lib/strings/str_util.cc]: `const size_t string_max_len` instead of `strsz`

**Review by drpngx — DISMISSED** [review]

**Comment by sb2nov** [comment]
@AKindyakov can you look at the merge conflict and the PR comments

**Comment by AKindyakov** [comment]
Yes, sure, I'll do it this night.

**Comment by rmlarsen** [comment]
I honestly don't have time to look at this right now, unassigning myself. @sb2nov could you re-assign, please?

**Comment by sb2nov** [comment]
@AKindyakov there is a merge conflict. Can you take a look.

**Comment by AKindyakov** [comment]
@sb2nov, it's done

**Comment by sb2nov** [comment]
Jenkins, test this please.

**Comment by AKindyakov** [comment]
Jenkins, test this, please.

**Comment by AKindyakov** [comment]
@sb2nov, I've fixed build and test problems. May I ask you to run full tests set for it?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by AKindyakov** [comment]
I've fixed makefile build and have fixed up dependencies lists in ``tensorflow/core``.
May I ask you to run full tests on it?

**Comment by AKindyakov** [comment]
Fixed problems with sanity check

**Comment by AKindyakov** [comment]
Can I ask someone to run the full test set for it?

**Comment by cwhipkey** [comment]
jenkins, test this please

File: tensorflow/core/BUILD
Comment (cwhipkey) [inline tensorflow/core/BUILD]: It's a little unfortunate that mobile builds will end up using it, when I assume they don't need the extra speed of parsing (since they will only be parsing a few attribute values, not large files of text).  Do you know how large the double-conversion library is when compiled for android?

(one option if the net change in binary size would be too large is to have the double conversion library only used by the op that needs it for speed).

File: tensorflow/core/lib/strings/numbers.cc
Comment (cwhipkey) [inline tensorflow/core/lib/strings/numbers.cc]: why this change?  (also, we could possibly use DoubleToStringConverter to implement all of DoubleToBuffer?)  (but see the question about android code size as well).

File: tensorflow/core/lib/strings/numbers_test.cc
Comment (cwhipkey) [inline tensorflow/core/lib/strings/numbers_test.cc]: maybe test case of leading and trailing whitespace as well

File: tensorflow/core/lib/strings/numbers.cc
Comment (cwhipkey) [inline tensorflow/core/lib/strings/numbers.cc]: do we need to support OCTALS?

File: tensorflow/core/BUILD
Comment (AKindyakov) [reply tensorflow/core/BUILD]: It's not so large actually. The size of the compiled static library is 486KiB on my laptop.

File: tensorflow/core/lib/strings/numbers.cc
Comment (AKindyakov) [reply tensorflow/core/lib/strings/numbers.cc]: I don't know. Should I remove it for the sake of nice little interface?

Comment (cwhipkey) [reply tensorflow/core/lib/strings/numbers.cc]: sgtm, since we didn't support before I'd rather not add support if we don't really need it.

File: tensorflow/core/lib/strings/numbers.cc
Comment (AKindyakov) [reply tensorflow/core/lib/strings/numbers.cc]: The previous implementation of DoubleToBuffer function was using old locale_independent_strtonum<double> which is replaced by double_conversion::StringToDoubleConverter with slightly differen interface.

But it is a good point to use only one function to convert str to double everywhere here. Consider double_conversion::StringToDoubleConverter only as hidden implementation details of safe_strto{d,f}.

This is a good idea to use double-conversion for {Double,Float}ToBuffer functions. I can do it here or in another pull request.

File: tensorflow/core/lib/strings/numbers_test.cc
Comment (AKindyakov) [reply tensorflow/core/lib/strings/numbers_test.cc]: Sure, I'll do it. Thank you.

File: tensorflow/core/BUILD
Comment (AKindyakov) [reply tensorflow/core/BUILD]: Stripped binary build with optimisation level 3 has size 77KiB. By the way, I can figure out what can we throw out from the library in the android build.

File: tensorflow/core/lib/strings/numbers.cc
Comment (AKindyakov) [reply tensorflow/core/lib/strings/numbers.cc]: ok, I'll remove it.

**Comment by AKindyakov** [comment]
@cwhipkey I fixed up ```CMake``` build, remove octal number support and add test cases with trailing and leading characters. Could you please run tests on it?

I was looking for a code in double conversion library which we can throw out in the android build. But this library is very solid and there is nothing to cut off.

**Comment by cwhipkey** [comment]
+ @petewarden  - pete, this will increase binary size some on  mobile (maybe not the full 77k, but it will probably be an increase).  What do you think we should do?   Some options are:

1. use ifdef to use the old code on mobile.  This clearly would cause no binary increase on mobile, but makes another code path that will be less tested and it's a bit unfortunate to have 2 code paths.

2. similar to 1., but do it in a way that can be tested off android.

3. add in the performant library and code only use by the kernels, but not in the string library.  This seems not great, since you would need to call the fast library from the kernels, instead of just using the string library.

4. test more fully and maybe it won't be a big change, or we can find savings elsewhere.

**Comment by samtaylor159** [comment]
Is there any update on this? This issue seems to be a current bottleneck in my code, it adds a huge overhead when reading my data.

**Comment by cwhipkey** [comment]
There is a pull request for it:
https://github.com/tensorflow/tensorflow/pull/12102

**Review by petewarden — APPROVED** [review]
Sorry for the delay! We may end up figuring out how to avoid the extra dependency on mobile as an option after this goes in, but I don't want that to block this change, so approving from my side.

**Review by cwhipkey — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
There are still problems with the Windows build, relevant error:

```
17:22:02 LINK : fatal error LNK1181: cannot open input file 'double_conversion\src\double_conversion\double-conversion\double-conversion.lib' [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\proto_text.vcxproj]
```

**Comment by martinwicke** [comment]
@mrry, do you know how to fix this?

**Comment by mrry** [comment]
At a guess, maybe the static library has a different filename on Windows? Maybe there's a missing `lib/` in the definition of `double_conversion_STATIC_LIBRARIES`?

**Comment by AKindyakov** [comment]
Is it me suppose to fix this?
I don't have any WIN machine for the test.

**Comment by drpngx** [comment]
@AKindyakov yes usually the committer ensures that everything builds, with our help. For this one, I pushed a commit that I think should fix it.

**Comment by drpngx** [comment]
@AKindyakov yes usually the committer ensures that everything builds, with our help. For this one, I pushed a commit that I think should fix it.

**Comment by drpngx** [comment]
@AKindyakov yes usually the committer ensures that everything builds, with our help. For this one, I pushed a commit that I think should fix it.

**Comment by AKindyakov** [comment]
@drpngx thanks a lot!

**Comment by drpngx** [comment]
Looks like there's a sycl breakage at `master`. Let's wait.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by AKindyakov** [comment]
@martinwicke, @drpngx have you any idea what is wrong with the build. There are no errors in build log only warnings which are not related to my changes at all.

**Comment by drpngx** [comment]
The mac test failed with a timeout.

The CMake failed with the same error as before: 
```
22:31:05 LINK : fatal error LNK1181: cannot open input file 'double_conversion\src\double_conversion\double-conversion\lib\double-conversion.lib' [C:\tf_jenkins\home\workspace\tensorflow-pr-win-cmake-py\cmake_build\proto_text.vcxproj]
```
The other errors don't appear to be related. Let's try again.

**Comment by AKindyakov** [comment]
@martinwicke, @drpngx, I believe I fixed up windows build. May I ask you to test it, please? (sorry for the delay - busy time)

**Comment by martinwicke** [comment]
Jenkins. Test this please.

**Comment by sb2nov** [comment]
@martinwicke is there a way to fix the CLA bot here. It was approved yes but the check seems to be pending.

**Comment by martinwicke** [comment]
It looks to me like it is stuck. I will override. I guess the XLA tests are known to be broken?

@willnorris FYI, not sure how to get it out of its funk.

**Comment by willnorris** [comment]
There's no obvious reason that I can see that would have caused this to get stuck, but I've manually approved it and will look into the cause.

**Comment by AKindyakov** [comment]
Thanks!

**Comment by caisq** [comment]
Just as a heads-up: This contribution is causing difficulties when the code is sync'ed to Google internal. We are considering reverting it first to unblock the code sync'ing, and maybe finding a way to make a similar change in internal.

cc @gunan

**Pull request merged**
